src/hotspot/share/gc/g1/g1YoungGCPostEvacuateTasks.cpp

```diff
@@ -354,6 +361,40 @@ class RedirtyLoggedCardTableEntryClosure : public G1CardTableEntryClosure {
   size_t num_dirtied()   const { return _num_dirtied; }
 };
 
+class G1PostEvacuateCollectionSetCleanupTask2::ClearRetainedRegionBitmaps : public G1AbstractSubTask {
+  G1EvacFailureRegions* _evac_failure_regions;
+  HeapRegionClaimer _claimer;
+
+  class ClearRetainedRegionBitmapsClosure : public HeapRegionClosure {
+  public:
+
+    bool do_heap_region(HeapRegion* r) override {
+      G1CollectedHeap::heap()->clear_bitmap_for_region(r, false /* update_tams */);
+      return false;
```

Comment:
I wonder if it's possible to assert `tams == bottom` here.

Reply:
No, after some thinking. G1 clears TAMSes during concurrent clearing of the bitmap. This marking phase is already concurrent to the mixed phase. We could reset TAMS anyway though.

Reply:
On second thought, yes bottom should be tams for regions in the collection set (of which retained regions are a subset).